src/models/ClassPotentialCPintheDark.cpp

```diff
@@ -1340,59 +1287,221 @@ void Class_Potential_CPintheDark::TripleHiggsCouplings()
+
+  // Indices of mass eigenstates for rotation from interaction to mass basis
+  pos_Gp = -1, pos_Gm = -1, pos_Hp = -1, pos_Hm = -1, pos_HSM = -1,
+  pos_G0 = -1, pos_h1 = -1, pos_h2 = -1, pos_h3 = -1;
 
-  for (size_t i = 0; i < NHiggs; i++)
-  {
-    // the rotation matrix is diagonal besides for the neutral dark scalars
-    if (std::abs(HiggsRot(i, 0)) > ZeroThreshold)
-      posGp = i;
-    else if (std::abs(HiggsRot(i, 1)) > ZeroThreshold)
-      posGm = i;
-    else if (std::abs(HiggsRot(i, 2)) > ZeroThreshold)
-      posHp = i;
-    else if (std::abs(HiggsRot(i, 3)) > ZeroThreshold)
-      posHm = i;
-    else if (std::abs(HiggsRot(i, 4)) > ZeroThreshold)
-      posHSM = i;
-    else if (std::abs(HiggsRot(i, 5)) > ZeroThreshold)
-      posG0 = i;
 
+  // basis = {rho1, eta1, rho2, eta2, zeta1, psi1, zeta2, psi2, rhoS}
+  // the rotation matrix is diagonal besides for the neutral dark scalars
+  for (std::size_t i = 0; i < NHiggs; i++) // mass base index i corresponds to mass vector sorted in ascending mass
+  {
+    if (std::abs(HiggsRot(i, pos_rho1)) > ZeroThreshold)
+    {
+      pos_Gp = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_eta1)) > ZeroThreshold)
+    {
+      pos_Gm = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_rho2)) > ZeroThreshold)
+    {
+      pos_Hp = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_eta2)) > ZeroThreshold)
+    {
+      pos_Hm = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_zeta1)) > ZeroThreshold)
+    {
+      pos_HSM = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_psi1)) > ZeroThreshold)
+    {
+      pos_G0 = i;
+    }
     // the neutral dark scalars mix
-    if ((std::abs(HiggsRot(i, 6)) + std::abs(HiggsRot(i, 7)) +
-         std::abs(HiggsRot(i, 8))) > ZeroThreshold)
+    else if ((std::abs(HiggsRot(i, pos_zeta2)) + std::abs(HiggsRot(i, pos_psi2)) +
+         std::abs(HiggsRot(i, pos_rhoS))) > ZeroThreshold)
     {
       // use that scalars are sorted by mass
-      if (posh1 == 0)
+      if (pos_h1 == -1)
+      {
+        pos_h1 = i;
+      }
+      else if (pos_h2 == -1)
       {
-        posh1 = i;
+        pos_h2 = i;
       }
       else
       {
-        if (posh2 == 0)
-        {
-          posh2 = i;
-        }
-        else
-        {
-          posh3 = i;
-        }
+        pos_h3 = i;
       }
     }
   }
 
-  // new rotation matrix with
-  MatrixXd HiggsRotSort(NHiggs, NHiggs);
+  // check if all position indices are set
+  if (pos_Gp == -1 or pos_Gm == -1 or pos_Hp == -1 or pos_Hm == -1 or
+      pos_HSM == -1 or pos_G0 == -1 or
+      pos_h1 == -1 or pos_h2 == -1 or pos_h3 == -1)
+  {
+    throw std::runtime_error("Error. Not all position indices are set.");
+  }
+
+  // check if all other elements of rotation matrix are zero
+  bool zero_element = false;
+  for (std::size_t i = 0; i < NHiggs; i++)
+  {
+    for (std::size_t j = 0; j < NHiggs; j++)
+    {
+      int ii = int(i);
+      int jj = int(j);
+      if (not((jj == pos_rho1  and ii == pos_Gp) or
```

Comment:
This piece of code is a bit confusing, what is it checking for? Can you refactor it a little bit?